### PR TITLE
New C++ API: Keep reference to solver object in non-solver objects.

### PR DESCRIPTION
**Pull request by aniemetz**
This is in preparation for adding guards to ensure that sort and term
arguments belong to the same solver.

@makaimann

### CODE REVIEW

File: src/api/cvc4cpp.cpp [hunk 1]
```diff
@@ -2169,28 +2192,34 @@ DatatypeConstructor Datatype::getConstructorForName(
...
-  return (*d_dtype)[index];
+  return DatatypeConstructor(d_solver, (*d_dtype)[index]);
 }
 
-Datatype::const_iterator::const_iterator(const CVC4::Datatype& dtype,
+Datatype::const_iterator::const_iterator(const Solver* slv,
+                                         const CVC4::Datatype& dtype,
                                          bool begin)
 {
+  d_solver = slv;
```
Comment (4tXJ7f) [inline src/api/cvc4cpp.cpp:2202]: Minor: Could be moved to initializer list.

File: src/parser/parser.cpp [hunk 2]
```diff
@@ -166,7 +161,7 @@ api::Sort Parser::getSort(const std::string& name)
 {
   checkDeclaration(name, CHECK_DECLARED, SYM_SORT);
   assert(isDeclared(name, SYM_SORT));
-  api::Sort t = api::Sort(d_symtab->lookupType(name));
+  api::Sort t = api::Sort(d_solver, d_symtab->lookupType(name));
```
Comment (4tXJ7f) [inline src/parser/parser.cpp:164]: Just a minor note on these kinds of assignments, we can actually just write `api::Sort t(d_solver, d_symtab->lookupType(name));` here. Probably out of scope for this PR but just something I noticed.

File: src/parser/tptp/Tptp.g [hunk 3]
```diff
@@ -1437,12 +1437,12 @@ tffLetTermBinding[std::vector<CVC4::api::Term> & bvlist,
                   CVC4::api::Term& lhs,
                   CVC4::api::Term& rhs]
   : term[lhs] EQUAL_TOK term[rhs]
-  {
-    PARSER_STATE->checkLetBinding(bvlist, lhs, rhs, false);
-    std::vector<api::Term> lchildren(++lhs.begin(), lhs.end());
-    rhs = MK_TERM(api::LAMBDA, MK_TERM(api::BOUND_VAR_LIST, lchildren), rhs);
-    lhs = api::Term(lhs.getExpr().getOperator());
-  }
+{
```
Comment (4tXJ7f) [inline src/parser/tptp/Tptp.g:1440]: This is very nit-picky but the indentation here is off: the curly brackets should align with the beginning of `term[lhs]` above.

Comment (aniemetz) [reply src/parser/tptp/Tptp.g:1440]: Clang-format screwed with this one, did we ever decide what to do to resolve this problem? Switch it off for the .g files?

File: src/api/cvc4cpp.cpp (hunk 1 above)
Comment (aniemetz) [reply src/api/cvc4cpp.cpp:2202]: Yeah, didn't move it because the rest is also initialized in the body.

File: src/parser/tptp/Tptp.g (hunk 3 above)
Comment (ajreynol) [reply src/parser/tptp/Tptp.g]: We don't have a good solution.  I think we decided to do our best effort to keep things close to clang-formatted everywhere, but don't require it if its a major effort.

**Pull request merged**